### PR TITLE
[SPARK-6638] [SQL] Improve performance of StringType in SQL

### PR DESCRIPTION
**Pull request by davies**
This PR change the internal representation for StringType from java.lang.String to UTF8String, which is implemented use Array[Byte](encoded in UTF-8).

This PR should not break any public API, Row.getString() will still return java.lang.String.

This is the first step of improve the performance of String in SQL.

cc @rxin

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala [hunk 1]
```diff
@@ -62,7 +62,11 @@ object IntegerLiteral {
   }
 }
 
-case class Literal(value: Any, dataType: DataType) extends LeafExpression {
+case class Literal(var value: Any, dataType: DataType) extends LeafExpression {
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala]: we should create a object Literal and add an apply function that converts string into UTF8String, rather than making this a var.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/dataTypes.scala [hunk 2]
```diff
@@ -386,6 +386,13 @@ class BinaryType private() extends NativeType with PrimitiveType {
       }
       x.length - y.length
     }
+    override def equiv(x: Array[Byte], y: Array[Byte]): Boolean = {
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/dataTypes.scala]: what is this used for?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/types/UTF8StringSuite.scala [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+      assert(UTF8String(str.getBytes("utf8")) == UTF8String(str))
+
+      assert(UTF8String(str).hashCode() == UTF8String(str.getBytes("utf8")).hashCode())
+    }
+
+    check("hello", 5)
+    check("世 界", 3)
+  }
+
+  test("contains, prefix and suffix") {
```
Comment (rxin) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/types/UTF8StringSuite.scala]: let's break these into 3 different test cases

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/UTF8String.scala [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+  }
+
+  override def hashCode(): Int = {
+    java.util.Arrays.hashCode(bytes)
+  }
+}
+
+private[sql] object UTF8String {
+  // number of tailing bytes in a UTF8 sequence for a code point
+  private[types] val bytesFromUTF8: Array[Int] = Array(1, 1, 1, 1, 1,
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/UTF8String.scala]: are there any links you can include for reference?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/GeneratedMutableEvaluationSuite.scala [hunk 5]
```diff
@@ -44,14 +43,6 @@ class GeneratedMutableEvaluationSuite extends ExpressionEvaluationSuite {
 
     val actual = plan(inputRow)
     val expectedRow = new GenericRow(Array[Any](expected))
-    if (actual.hashCode() != expectedRow.hashCode()) {
```
Comment (davies) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/GeneratedMutableEvaluationSuite.scala]: The hashCode of UTF8String is different than String

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala [hunk 6]
```diff
@@ -62,7 +75,10 @@ object IntegerLiteral {
   }
 }
 
-case class Literal(value: Any, dataType: DataType) extends LeafExpression {
+case class Literal (var value: Any, dataType: DataType) extends LeafExpression {
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala]: There are 400-500 places using this constructor, I'd like to leave the refactor out for another PR.

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala]: I think you are right. Can we submit a separate PR to avoid using the constructor directly? Then we can merge this one.

**Comment by davies** [comment]
@rxin @marmbrus This PR is ready to review

**Comment by rxin** [comment]
Can we open a new PR and close this one to get rid of the all the Jenkins messages?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala [hunk 7]
```diff
@@ -80,6 +82,24 @@ trait ScalaReflection {
     case (other, _) => other
   }
 
+  /**
+   *  Converts Scala objects to catalyst rows / types.
+   *  Note: This should be called before do evaluation on Row
+   *        (It does not support UDT)
+   */
+  def convertToCatalyst(a: Any): Any = a match {
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala]: was this function somewhere before?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificMutableRow.scala [hunk 8]
```diff
@@ -230,13 +247,15 @@ final class SpecificMutableRow(val values: Array[MutableValue]) extends MutableR
     new GenericRow(newValues)
   }
 
-  override def update(ordinal: Int, value: Any): Unit = {
-    if (value == null) setNullAt(ordinal) else values(ordinal).update(value)
+  override def update(ordinal: Int, value: Any): Unit = value match {
+    case null => setNullAt(ordinal)
+    case s: String => update(ordinal, UTF8String(s))
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificMutableRow.scala]: do we ever have a string here? since this is used internally by spark sql, it seems wrong to expect a string

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala (hunk 7 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala]: No, we have a similar one: `convertToCatalyst (a: Any, dt: DataType)`

**Comment by davies** [comment]
@rxin I will create a new PR after fixing the rest tests.

**Comment by davies** [comment]
Close this one to get rid of all the jenkins comments.
